dali/operators/reader/loader/video/frames_decoder.cc

```diff
@@ -0,0 +1,259 @@
+      Width(), 
+      Height(), 
+      av_state_->codec_ctx_->pix_fmt, 
+      Width(), 
+      Height(), 
+      AV_PIX_FMT_RGB24, 
+      SWS_BILINEAR, 
+      nullptr, 
+      nullptr, 
+      nullptr);
+    DALI_ENFORCE(av_state_->sws_ctx_, "Could not create sw context");
+  }
+}
+
+bool FramesDecoder::ReadRegularFrame(uint8_t *data, bool copy_to_output) {
+  int ret = -1;
+  while (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
+    if (av_state_->packet_->stream_index != av_state_->stream_id_) {
+      continue;
+    }
+
+    ret = avcodec_send_packet(av_state_->codec_ctx_, av_state_->packet_);
+    DALI_ENFORCE(
+      ret >= 0,
+      make_string("Failed to send packet to decoder: ", detail::av_error_string(ret)));
+
+    ret = avcodec_receive_frame(av_state_->codec_ctx_, av_state_->frame_);
+
+    if (ret == AVERROR(EAGAIN)) {
+      continue;
+    }
+
+    if (ret == AVERROR_EOF) {
+      break;
+    }
+
+    if (!copy_to_output) {
+      return true;
+    }
+
+    LazyInitSwContext();
+
+    CopyToOutput(data);
+    return true;
+  }
+
+  ret = avcodec_send_packet(av_state_->codec_ctx_, nullptr);
+  DALI_ENFORCE(
+    ret >= 0,
+    make_string("Failed to send packet to decoder: ", detail::av_error_string(ret)));
+  flush_state_ = true;
+
+  return false;
+}
+
+void FramesDecoder::Reset() {
+  int ret = av_seek_frame(av_state_->ctx_, av_state_->stream_id_, 0, AVSEEK_FLAG_FRAME);
+  DALI_ENFORCE(
+    ret >= 0,
+    make_string(
+      "Could not seek to the first frame of video ",
+      filename_,
+      "due to",
+      detail::av_error_string(ret)));
+  avcodec_flush_buffers(av_state_->codec_ctx_);
+}
+
+void FramesDecoder::SeekFrame(int frame_id) {
+  // TODO(awolant): Optimize seeking: 
+  //  - when we seek next frame,
+  //  - when we seek frame with the same keyframe as the current frame
+  //  - for CFR, when we know pts, but don't know keyframes 
+
+  DALI_ENFORCE(
+    frame_id >= 0 && frame_id < NumFrames(),
+    make_string("Invalid seek frame id. frame_id = ", frame_id, ", num_frames = ", NumFrames()));
+
+  auto &frame_entry = index_[frame_id];
+  int keyframe_id = frame_entry.last_keyframe_id;
+  auto &keyframe_entry = index_[keyframe_id];
+
+  // Seeking clears av buffers, so reset flush state info
+  if (flush_state_) {
+    flush_state_ = false;
+  }
+ 
+  int ret = av_seek_frame(av_state_->ctx_, av_state_->stream_id_, keyframe_entry.pts, AVSEEK_FLAG_FRAME);
+  DALI_ENFORCE(
+    ret >= 0,
+    make_string(
+      "Failed to seek to frame ",
+      frame_id,
+      "with keyframe",
+      keyframe_id,
+      "in video ",
+      filename_,
+      "due to ",
+      detail::av_error_string(ret)));
+
+  avcodec_flush_buffers(av_state_->codec_ctx_);
+
+  for (int i = 0; i < (frame_id - keyframe_id); ++i) {
+    ReadNextFrame(nullptr, false);
+  }
+}
+
+bool FramesDecoder::ReadFlushFrame(uint8_t *data, bool copy_to_output) {
+  if (avcodec_receive_frame(av_state_->codec_ctx_, av_state_->frame_) < 0) {
```

Comment:
Is it possible to get fatal error that is not expected here and when we should just throw?